src/test/kotlin/org/rustSlowTests/lang/resolve/CargoProjectResolveTest.kt

```diff
@@ -947,6 +947,79 @@ class CargoProjectResolveTest : RsWithToolchainTestBase() {
         prj.checkReferenceIsResolved<RsPath>("project_2/src/main.rs")
     }
 
+    fun `test cargo features namespaced dependencies and weak dependency features`() = buildProject {
+        toml("Cargo.toml", """
+            [package]
+            name = "hello"
+            version = "0.1.0"
+
+            [features]
+            feature_foo = ["dep:foo"]
+            feature_bar = ["dep:bar"]
+            feature_baz = ["foo?/baz"]
+
+            [dependencies]
+            foo = { path = "./foo", optional = true }
+            bar = { path = "./bar", optional = true }
+        """)
+
+        dir("src") {
+            rust("main.rs", """
+                fn main() {
+                    foo::foo_fn();
+                }       //^
+            """)
+        }
+
+        dir("foo") {
+            toml("Cargo.toml", """
+                [package]
+                name = "foo"
+                version = "0.1.0"
+
+                [features]
+                baz = []
+            """)
+
+            dir("src") {
+                rust("lib.rs", """
+                    #[cfg(feature = "baz")]
+                    pub fn foo_fn() {}
+                """)
+            }
+        }
+
+        dir("bar") {
+            toml("Cargo.toml", """
+                [package]
+                name = "bar"
+                version = "0.1.0"
+            """)
+
+            dir("src") {
+                rust("lib.rs", "pub fn bar_fn() {}")
+            }
+        }
+    }.run {
+        val pkg = project.cargoProjects.singlePackage("hello")
+        assertEquals(setOf("feature_foo", "feature_bar", "feature_baz"), pkg.featureState.keys)
+
+        pkg.checkFeatureEnabled("feature_foo")
+        pkg.checkFeatureEnabled("feature_bar")
+        pkg.checkFeatureEnabled("feature_baz")
+        project.cargoProjects.singlePackage("foo").checkFeatureEnabled("baz")
+
+        checkReferenceIsResolved<RsPath>("src/main.rs")
+
+        project.cargoProjects.disableCargoFeature("hello", "feature_foo")
+        project.cargoProjects.disableCargoFeature("hello", "feature_bar")
+        project.cargoProjects.disableCargoFeature("hello", "feature_baz")
+
+        project.cargoProjects.singlePackage("foo").checkFeatureDisabled("baz")
+
```

Comment:
If I understand right, we could also check this scenario:
```Kotlin
project.cargoProjects.disableCargoFeature("hello", "feature_foo")
project.cargoProjects.enableCargoFeature("hello", "feature_baz")

fooPackage.checkFeatureDisabled("baz")
```
It looks like currently the feature `baz` will be enabled in that case. But referring to the example from the [blog post](https://blog.rust-lang.org/2022/04/07/Rust-1.60.0.html#new-syntax-for-cargo-features):
```Toml
[dependencies]
serde = { version = "1.0.133", optional = true }
rgb = { version = "0.8.25", optional = true }

[features]
serde = ["dep:serde", "rgb?/serde"]
```
> In this example, enabling the serde feature will enable the serde dependency. It will also enable the serde feature for the rgb dependency, but only if something else has enabled the rgb dependency.

Reply:
> It looks like currently the feature `baz` will be enabled in that case

No, it will be enabled because there isn't a feature named `baz`. If you use `dep:foo` or `foo?/baz` syntax, cargo does not add an implicit feature for the optional dependency, so there just no feature named `baz`, just `feature_foo`, `feature_bar` and `feature_baz`. `baz` is not a feature and hence there isn't a way to check whether it is enabled or not. So `fooPackage.checkFeatureDisabled("baz")` just does something weird and fails, so looks like a wrong test failure message confused you

Reply:
Oh you're right, `baz` must be disabled in this case

Reply:
It seems quite hard to fix. Maybe we can merge the PR with this bug, create an issue about it and fix it separately?

Reply:
Opened an issue #9476